[torch.compile] fast inductor

youkaichao opened the pull request:
directly bypass aot-autograd and inductor, and load from cache

youkaichao committed:
show draft

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into fast_inductor

youkaichao committed:
polish

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
change read

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
use InductorHashCache class

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename key to hash_str

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
comment

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
python-style

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix serialize

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/config.py:
```diff
@@ -2212,6 +2215,53 @@ class CompilationLevel:
     PIECEWISE = 3
 
 
+class InductorHashCache:
```
tried to place this class into `vllm.compilation.backends` , but then needs to be lazily imported, and `pydantic` will complain.

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
now, even if we compile for all sizes, the compilation time becomes negligible:

```bash
$ vllm serve meta-llama/Meta-Llama-3-8B --disable-log-requests -O "{'level': 3, 'candidate_compile_sizes': [$(seq -s, 1 1 256)]}"
INFO 12-11 14:04:25 backends.py:363] Dynamo bytecode transform time: 4.59 s
INFO 12-11 14:04:28 backends.py:155] Compiling a graph for general shape takes 2.77 s
INFO 12-11 14:04:32 backends.py:158] Compiling a graph for shape 256 takes 0.48 s
INFO 12-11 14:04:33 backends.py:158] Compiling a graph for shape 248 takes 0.48 s
INFO 12-11 14:04:34 backends.py:158] Compiling a graph for shape 240 takes 0.49 s
INFO 12-11 14:04:35 backends.py:158] Compiling a graph for shape 232 takes 0.37 s
INFO 12-11 14:04:36 backends.py:158] Compiling a graph for shape 224 takes 0.38 s
INFO 12-11 14:04:37 backends.py:158] Compiling a graph for shape 216 takes 0.47 s
INFO 12-11 14:04:37 backends.py:158] Compiling a graph for shape 208 takes 0.32 s
INFO 12-11 14:04:38 backends.py:158] Compiling a graph for shape 200 takes 0.47 s
INFO 12-11 14:04:39 backends.py:158] Compiling a graph for shape 192 takes 0.39 s
INFO 12-11 14:04:39 backends.py:158] Compiling a graph for shape 184 takes 0.34 s
INFO 12-11 14:04:40 backends.py:158] Compiling a graph for shape 176 takes 0.36 s
INFO 12-11 14:04:41 backends.py:158] Compiling a graph for shape 168 takes 0.49 s
INFO 12-11 14:04:42 backends.py:158] Compiling a graph for shape 160 takes 0.52 s
INFO 12-11 14:04:43 backends.py:158] Compiling a graph for shape 152 takes 0.47 s
INFO 12-11 14:04:44 backends.py:158] Compiling a graph for shape 144 takes 0.41 s
INFO 12-11 14:04:44 backends.py:158] Compiling a graph for shape 136 takes 0.33 s
INFO 12-11 14:04:45 backends.py:158] Compiling a graph for shape 128 takes 0.53 s
INFO 12-11 14:04:46 backends.py:158] Compiling a graph for shape 120 takes 0.33 s
INFO 12-11 14:04:47 backends.py:158] Compiling a graph for shape 112 takes 0.48 s
INFO 12-11 14:04:48 backends.py:158] Compiling a graph for shape 104 takes 0.51 s
INFO 12-11 14:04:48 backends.py:158] Compiling a graph for shape 96 takes 0.53 s
INFO 12-11 14:04:49 backends.py:158] Compiling a graph for shape 88 takes 0.54 s
INFO 12-11 14:04:50 backends.py:158] Compiling a graph for shape 80 takes 0.52 s
INFO 12-11 14:04:51 backends.py:158] Compiling a graph for shape 72 takes 0.59 s
INFO 12-11 14:04:52 backends.py:158] Compiling a graph for shape 64 takes 0.57 s
INFO 12-11 14:04:53 backends.py:158] Compiling a graph for shape 56 takes 0.51 s
INFO 12-11 14:04:54 backends.py:158] Compiling a graph for shape 48 takes 0.42 s
INFO 12-11 14:04:55 backends.py:158] Compiling a graph for shape 40 takes 0.56 s
INFO 12-11 14:04:56 backends.py:158] Compiling a graph for shape 32 takes 0.44 s
INFO 12-11 14:04:57 backends.py:158] Compiling a graph for shape 24 takes 0.47 s
INFO 12-11 14:04:57 backends.py:158] Compiling a graph for shape 16 takes 0.47 s
INFO 12-11 14:04:58 backends.py:158] Compiling a graph for shape 8 takes 0.47 s
INFO 12-11 14:04:59 backends.py:158] Compiling a graph for shape 4 takes 0.52 s
INFO 12-11 14:05:00 backends.py:158] Compiling a graph for shape 2 takes 0.51 s
INFO 12-11 14:05:02 backends.py:158] Compiling a graph for shape 1 takes 1.43 s
INFO 12-11 14:05:02 monitor.py:31] torch.compile takes 24.54 s in total
```

youkaichao commented:
cc @ProExpertProg @bnellnm

youkaichao commented:
throughput benchmark:

```bash
# run baseline first, to find out the number of scheduler steps to keep gpus busy
$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B
Throughput: 30.52 requests/s, 12620.07 total tokens/s, 6052.88 output tokens/s

$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 8
Throughput: 43.57 requests/s, 18018.38 total tokens/s, 8642.04 output tokens/s

$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 10
Throughput: 44.04 requests/s, 18212.79 total tokens/s, 8735.28 output tokens/s

$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 12
Throughput: 44.65 requests/s, 18464.38 total tokens/s, 8855.95 output tokens/s

$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 16
Throughput: 45.03 requests/s, 18622.54 total tokens/s, 8931.81 output tokens/s

$ python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 20
Throughput: 44.18 requests/s, 18269.12 total tokens/s, 8762.30 output tokens/s

# the best number of scheduler step is 16, run this setting with `torch.compile`
python benchmarks/benchmark_throughput.py --dataset ShareGPT_V3_unfiltered_cleaned_split.json --model meta-llama/Meta-Llama-3-8B --num-scheduler-steps 16 -O "{'level': 3, 'candidate_compile_sizes': [$(seq -s, 1 1 256)]}"
Throughput: 46.93 requests/s, 19405.53 total tokens/s, 9307.35 output tokens/s
```

Now we don't need to profile the batchsize distribution anymore. We can just compile for all sizes.

youkaichao committed:
fix high-order ops

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix splitting ops

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
I think there might be something wrong inside Inductor when I try llama 3 70B model (with tp 4). Even when I directly cache the graph hash, every shape takes 90 seconds to load the graph.

I will stop here as the results on llama 3 8B are quite good, compilation time is greatly reduced. The llama 3 70B case should be a bug to fix in the future.

youkaichao commented on vllm/config.py:
```diff
@@ -2649,6 +2726,29 @@ def __post_init__(self):
+        if self.model_config is not None and \
+            not self.compilation_config.cache_dir:
+            # generate a cache directory based on the model information
+            # TODO: consider more factors that will affect model forward,
```
I think I missed some quantization args that can affect model execution, but I don't know how to pull out all factors that affect quantization.

tlrmchlsmth replied:
vLLM version? We can add the git SHA to the key

tlrmchlsmth replied:
This is going to be a large source of potential bugs so definitely should be careful here. Most quantization related stuff from NM goes in the `model_config` but there's a lot of arguments to `LLM` that can affect things like `dtype` and `quantization`. Are these in the key already?

youkaichao replied:
not yet. that's why I want to ask for reviews.

one direction is we consider all factors affecting compilation, and we can use compilation cache by default.

another approach is we don't cache by default, but tell user the cache directory, and users can specify the cache directory if they know nothing changed.

which one would you prefer?

ProExpertProg replied:
I think we should always check the known factors when we cache, and expose an accessible switch for enabling/disabling caching. And then it's less important whether it's on by default or not. And for that decision @robertgshaw2-neuralmagic should chime in.

youkaichao replied:
I added more factors to consider in [2a7f729](https://github.com/vllm-project/vllm/pull/11108/commits/2a7f7291a1fc8e5dbc07387169fbad7ebcb82f2a) . Let me know if I miss anything.

youkaichao commented on vllm/config.py:
```diff
@@ -2414,6 +2481,16 @@ def init_backend(self, vllm_config: "VllmConfig") -> Union[str, Callable]:
+        self.cache_dir = os.path.join(
+            self.cache_dir, f"rank_{vllm_config.parallel_config.rank}")
+        os.makedirs(self.cache_dir, exist_ok=True)
+        self.inductor_hash_cache_path = os.path.join(self.cache_dir,
```
it would be better if we also save a serialized form of the config, but we need to design the serialized format.

youkaichao replied:
it is serializable, but i want a human-readable form, so that we can also manually check the config.

ProExpertProg commented on vllm/compilation/backends.py:
```diff
@@ -55,9 +93,71 @@ def wrap_inductor(graph,
+
+        def _check_can_cache(*args, **kwargs):
+            # no error means it can be cached
+            # vLLM computation graph can always be cached
```
I think this might be too liberal. Why can't we just trust Inductor to decide if we can cache or not? Maybe this should respect the `compile_config.inductor_config["force_disable_caches"]`? Or more likely we should have our own config variable for disabling the whole process

youkaichao replied:
> Why can't we just trust Inductor to decide if we can cache or not?

this is not a flag for inductor to cache the graph in the disk, it's about inductor will ever use cache for the graph or not. it is not affected by any inductor configs.

I added the explanation at [4f1c4a0](https://github.com/vllm-project/vllm/pull/11108/commits/4f1c4a016a9406cd928622f5326d463487fa8961) . let me know if it is clear now.

ProExpertProg commented on vllm/compilation/backends.py:
```diff
@@ -457,6 +557,12 @@ def __call__(self, *args) -> Any:
             # finished compilations for all required shapes
             if self.is_last_graph and not self.to_be_compiled_sizes:
+
+                # save the hash of the inductor graph for the next run
+                with open(self.compilation_config.inductor_hash_cache_path,
+                          "w") as f:
+                    f.write(self.compilation_config.inductor_hash_cache.
+                            serialize())
```
This should live inside `InductorHashCache`, or if you want to keep that class a pure data object, perhaps an `InductorHashCacheManager`

youkaichao replied:
good point! fixed in [955989c](https://github.com/vllm-project/vllm/pull/11108/commits/955989c3492388b6b32d9ba10a8e4484a07ec700)

ProExpertProg reviewed (commented):
Two main notes:
- Figuring out what the "key" is for the cache should be a method on config, likely on each sub-config as well. That way it's clear when developers modify config they might need to modify the key as well.
- I think `InductorHashCache` is a nice abstraction but it shouldn't live inside config (both file- and structure-wise). It should also take on more responsibilities, I think we can make it cleaner, so that `wrap_inductor` doesn't contain caching logic - it just calls the appropriate methods on the cache (or cache manager) object

youkaichao committed:
move file writing inside InductorHashCache

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/compilation/backends.py:
```diff
@@ -55,9 +93,71 @@ def wrap_inductor(graph,
+        # it to get the hash of the compiled graph directly.
+        from torch._inductor.codecache import compiled_fx_graph_hash
+
+        def mocked_compiled_fx_graph_hash(*args, **kwargs):
```
renamed to `hijack_compiled_fx_graph_hash` , does it sound good to you?

youkaichao committed:
give error message

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/compilation/backends.py:
```diff
@@ -55,9 +93,71 @@ def wrap_inductor(graph,
+        with patch("torch._inductor.codecache.FxGraphCache._get_shape_env",
+                   lambda *args, **kwargs: AlwaysHitShapeEnv()):
+            inductor_compiled_graph = FxGraphCache._lookup_graph(
+                hash_str, example_inputs, True, False)
```
good point, added in [37d744d](https://github.com/vllm-project/vllm/pull/11108/commits/37d744d3028fd47c654acf96d46e18b03845d56e)

youkaichao commented on vllm/config.py:
```diff
@@ -2212,6 +2215,53 @@ class CompilationLevel:
 
+class InductorHashCache:
+    """
+    Disk format: a Python list of tuples, each tuple is
```
Json does not allow integer as dict key.

I use this format mainly for readability, but if you have another solution, that also works. It's basically `Dict[Optional[int], Dict[int, str]]` .

youkaichao replied:
explained in [28b98fb](https://github.com/vllm-project/vllm/pull/11108/commits/28b98fb394d924b0e510cc1f65cf67060dced988)

youkaichao committed:
move to another file

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
> I think InductorHashCache is a nice abstraction but it shouldn't live inside config (both file- and structure-wise).

@ProExpertProg I agree. I moved `InductorHashCache` to the `backends.py` now. For the structure-wise, I think we should add some `VllmState` in the future, instead of adding all fields in the config. But that needs a large refactor though, beyond the scope of this current PR.

youkaichao committed:
add comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add more factors to consider

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/config.py:
```diff
@@ -2649,6 +2726,29 @@ def __post_init__(self):
+            cache_dir = os.path.join(envs.VLLM_CACHE_ROOT,
+                                     "torch_compile_cache", hash_str)
+            os.makedirs(cache_dir, exist_ok=True)
+            self.compilation_config.cache_dir = cache_dir
```
added in [2a7f729](https://github.com/vllm-project/vllm/pull/11108/commits/2a7f7291a1fc8e5dbc07387169fbad7ebcb82f2a)

youkaichao replied:
some configs should not affect the model execution, like scheduler config.

youkaichao commented:
I think I should addressed most comments now, let me know if you have any other concerns @ProExpertProg @tlrmchlsmth 

One thing I'm not sure, is whether I should redirect inductor code cache and triton cache to our vllm cache directory or not.

Upside1: Sometimes (rarely) I find it is possible to have hash collision. Because we directly load fx graph from cache here, we don't have any check for hash collision. Redirecting inductor code cache and triton cache to our vllm cache directory is better for soundness.

Upside2: the vllm cache directory becomes self-contained, users only need to copy the whole directory for fast deployment.

Downside: different models cannot share max-autotune cache (mainly for gemm performance tuning). The code start time will be longer.

ProExpertProg commented:
> Downside: different models cannot share max-autotune cache (mainly for gemm performance tuning). The code start time will be longer.
Why can't models share the autotune cache anymore?

youkaichao commented:
@ProExpertProg I plan to set `TORCHINDUCTOR_CACHE_DIR` to the cache directory we generate, this environment variable controls all of inductor cache.

tlrmchlsmth commented on vllm/config.py:
```diff
@@ -2649,11 +2702,52 @@ def __post_init__(self):
+        torch_factors = torch_key()
+        factors.append(torch_factors)
+
+        # summarize vllm config
```
There's a few entries in the SchedulerConfig that could affect the computation graph:
* max_num_batched_tokens
* max_num_seqs
* max_model_len

Not sure if they do currently. But I could see us making decisions that affect the compute graph based on them. 

One thing I have in mind here: When constructing the Flux kernels in https://github.com/vllm-project/vllm/pull/9886 need to supply a maximum `num_tokens`. We could use these for that purpose.

bnellnm replied:
It's not in the PR yet but I am using `max_num_batched_tokens` for the flux rewrite.

tlrmchlsmth commented on vllm/config.py:
```diff
@@ -2354,12 +2371,26 @@ def model_post_init(self, __context: Any) -> None:
     enabled_custom_ops: Counter[str] = PrivateAttr
     disabled_custom_ops: Counter[str] = PrivateAttr
     compilation_time: float = PrivateAttr
+    # should be InductorHashCache , but Pydantic does not support it
```
```suggestion
    # should be InductorHashCache, but Pydantic does not support it
```

tlrmchlsmth commented on vllm/config.py:
```diff
@@ -316,6 +317,15 @@ def __init__(
 
+    def compute_hash(self):
+        # summarize all the factors that affect the model compulation and
+        # hence model compilation, and use them to generate a hash
```
These look like they could plausibly affect the compute graph.

`revision` will affect the model that's loaded from HF, and rope scaling and theta could be baked in as constants. `quantization_param_path` is used for kv cache quantization
```
revision
code_revision
rope_scaling
rope_theta
max_model_len
quantization_param_path
max_logprobs
disable_sliding_window
```

youkaichao replied:
why would `max_logprobs` affect the computation?

I don't think `disable_sliding_window` affect the computation.

for rope-related field, I agree we should add it to the factors.

we hash the fx graph, as long as the config does not change the fx graph's structure, then it is safe to reuse the compilation.

therefore, anything inside the attention layer does not matter. the attention op is wrapped as a custom op.

tlrmchlsmth commented on vllm/config.py:
```diff
@@ -1063,6 +1073,12 @@ class ParallelConfig:
     rank: int = 0
 
+    def compute_hash(self):
+        factors: List[Any] = []
```
`disable_custom_all_reduce`?

And what about `worker_cls`?

youkaichao replied:
all reduce is wrapped as a custom op, and `disable_custom_all_reduce` only affects the internal of the custom op, it does not affect the fx graph.

worker class does not affect fx graph either. the model determines the fx graph.

tlrmchlsmth reviewed (commented):
I think we should make the inductor cache opt-in for users. Explicit control over use of the cache will help a lot with the potential errors.

Along similar lines, we could hash everything in the configs instead of only hashing certain member variables. I think a lot of these could affect the compute graph in unexpected ways.

youkaichao commented:
> we could hash everything in the configs instead of only hashing certain member variables. I think a lot of these could affect the compute graph in unexpected ways.

there's trade-off here. if we hash everything, then users need to recompile for some unnecessary changes, e.g. `--num-scheduler-steps` . I would still vote for only hashing certain variables we know they will affect the computation.

> I think we should make the inductor cache opt-in for users.

If we hash lots of the variables, do you think it makes sense to make it on by default, and add a flag to disable it? I'm thinking about the end-user experience, compilation time is really important for user experience, and it would be great if we can turn it on by default.

youkaichao committed:
Update vllm/config.py

Co-authored-by: Tyler Michael Smith <tyler@neuralmagic.com>

youkaichao committed:
typo

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/config.py:
```diff
@@ -316,6 +317,15 @@ def __init__(
         self._verify_bnb_config()
 
+    def compute_hash(self):
+        # summarize all the factors that affect the model compulation and
```
thanks, fixed in [75cf1f5](https://github.com/vllm-project/vllm/pull/11108/commits/75cf1f5c7a61043980c2048e357cc122e41cfcf4)

youkaichao committed:
Merge branch 'main' into fast_inductor

tlrmchlsmth commented:
> I would still vote for only hashing certain variables we know they will affect the computation.

Maybe we can hash everything in the config classes by default, but exclude certain items that we know are safe?

> If we hash lots of the variables, do you think it makes sense to make it on by default, and add a flag to disable it?

If we're confident that we're hashing everything relevant then, I agree it makes sense to be opt-out.

youkaichao committed:
merge conflict

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
consider all factors

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add vllm version

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
bugfix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add disable

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
redirect inductor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add more logging

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add more logging

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix shape 1

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/envs.py:
```diff
@@ -71,6 +71,7 @@
     VLLM_USE_V1: bool = False
     VLLM_ENABLE_V1_MULTIPROCESSING: bool = False
     VLLM_LOG_BATCHSIZE_INTERVAL: float = -1
+    VLLM_DISABLE_COMPILE_CACHE: bool = False
```
added the flag to disable the compile cache. the compile cache is used by default.

youkaichao commented on vllm/compilation/backends.py:
```diff
@@ -21,6 +25,126 @@
+                                            "inductor_hash_cache.py")
+        if disabled:
+            return
+        # set flags so that Inductor and Triton store their cache
+        # in the cache_dir, then users only need to copy the cache_dir
+        # to another machine to reuse the cache.
+        inductor_cache = os.path.join(cache_dir, "inductor_cache")
+        os.makedirs(inductor_cache, exist_ok=True)
+        os.environ["TORCHINDUCTOR_CACHE_DIR"] = inductor_cache
+        triton_cache = os.path.join(cache_dir, "triton_cache")
+        os.makedirs(triton_cache, exist_ok=True)
+        os.environ["TRITON_CACHE_DIR"] = triton_cache
```
redirect inductor/triton cache to the vllm cache location.

youkaichao commented:
In the future, I plan to dump more information in the cache directory, including:
- the transformed bytecode from Dynamo
- the human-readable fx graph after every transform
- the human-readable fx graph that inductor compiles
- the human-readable executable python file inductor finally run

the goal is to make `torch.compile` crystally clear, we can know the exact correspondence of the compiled code and our original code, so that we can easily debug any potential correctness / performance issues.

ProExpertProg reviewed (approved):
Agree with Tyler's comments about config caching, otherwise LGTM!

youkaichao committed:
fix tests

Signed-off-by: youkaichao <youkaichao@gmail.com>

tlrmchlsmth commented on vllm/config.py:
```diff
@@ -2365,12 +2585,33 @@ def model_post_init(self, __context: Any) -> None:
+        factors.append(self.custom_ops)
+        factors.append(self.splitting_ops)
+        factors.append(self.use_inductor)
+        factors.append(self.inductor_compile_config)
+        factors.append(self.inductor_passes)
```
What about `candidate_compile_sizes`?

youkaichao replied:
they do not affect the computation graph.

say we compile for `candidate_compile_sizes = [1, 2, 4]` , and then run again with `candidate_compile_sizes = [1, 2, 4, 8]` , we want them to share the same cache directory, so that we can directly load the cache for `[1, 2, 4]` , and only compile for a new shape `8` .

youkaichao committed:
fix inductor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into fast_inductor

tlrmchlsmth commented on vllm/compilation/backends.py:
```diff
@@ -55,9 +175,84 @@ def wrap_inductor(graph,
+        # we compiled this graph before
+        # so we can directly lookup the compiled graph via hash
+        hash_str = cache_data[(runtime_shape, graph_index)]
+        logger.debug(
+            "directly lookup the %s-th graph for shape %s via hash %s",
+            graph_index, str(runtime_shape), hash_str)
```
Could you add some info-level logging as well so we always know whether caching is/isn't being used from looking at the logs?

Can use `print_info_once` so it doesn't print for every layer
```
print_info_once("Using previously compiled compute graph for shape %s", hash_str)
```

youkaichao replied:
added in [c4478c8](https://github.com/vllm-project/vllm/pull/11108/commits/c4478c8a4c67251fb6475986bd6bf208371aae7a)

tlrmchlsmth commented on vllm/compilation/backends.py:
```diff
@@ -55,9 +175,84 @@ def wrap_inductor(graph,
+        # compiled_fx_graph_hash will only be called once, and we can hook
+        # it to get the hash of the compiled graph directly.
+        from torch._inductor.codecache import compiled_fx_graph_hash
+
```
Then similarly, we can print something if nothing is found as well. Assuming we'd never print both in ordinary circumstances - is that right?

```
print_info_once("Did not find previously compiled compute graph for shape %s". Compiling.", hash_str)
```

youkaichao replied:
print it by default now, let's see how users like it.

tlrmchlsmth reviewed (approved):
Works for me now! Some suggestions on comments and logging, but LGTM otherwise

youkaichao committed:
add warning

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/config.py:
```diff
@@ -158,6 +159,26 @@ class ModelConfig:
+        graph from input ids/embeddings to the final hidden states,
+        excluding anything before input ids/embeddings and after
+        the final hidden states.
+        """
```
fixed in [40f1355](https://github.com/vllm-project/vllm/pull/11108/commits/40f1355c86f71549c48c035a184936af2f069034)

youkaichao committed:
add info logging

Signed-off-by: youkaichao <youkaichao@gmail.com>

zou3519 commented:
> I think there might be something wrong inside Inductor when I try llama 3 70B model (with tp 4). Even when I directly cache the graph hash, every shape takes 90 seconds to load the graph.
> I will stop here as the results on llama 3 8B are quite good, compilation time is greatly reduced. The llama 3 70B case should be a bug to fix in the future.

@youkaichao is this still a problem?

youkaichao commented:
> > I think there might be something wrong inside Inductor when I try llama 3 70B model (with tp 4). Even when I directly cache the graph hash, every shape takes 90 seconds to load the graph.
> > I will stop here as the results on llama 3 8B are quite good, compilation time is greatly reduced. The llama 3 70B case should be a bug to fix in the future.
> 
> @youkaichao is this still a problem?

didn't check recently. we need to check and profile the start time.
